dnscache.go

```diff
@@ -0,0 +1,130 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+type DNSCache struct {
+	resolver *net.Resolver
+	mutex    sync.Mutex
+	size     int
+	duration time.Duration
+	entries  map[string]*dnsCacheEntry
+}
+
+func NewDNSCache(size int, duration time.Duration) *DNSCache {
+	return &DNSCache{
+		resolver: net.DefaultResolver,
```

Comment:
Please can you add some unit tests for the `DNSCache` e.g replacing `resolver *net.Resolver` with `resolver netResolver` where:
```go
type netResolver interface {
    LookupIPAddr(ctx, name)
}
```
which allows us to stub out the network operations. Similarly, you'll need to do this for `dialer := net.Dialer{}` and `DialContext`.

Reply:
Done - added tests.